src/base/tristatebool.cpp

```diff
@@ -28,33 +28,67 @@
 
 #include "tristatebool.h"
 
-TriStateBool::TriStateBool()
-    : m_value(Undefined)
-{
-}
+#include <QDebug>
+#include <QJsonValue>
+
+const TriStateBool TriStateBool::Undefined(-1);
+const TriStateBool TriStateBool::False(0);
+const TriStateBool TriStateBool::True(1);
 
 TriStateBool::TriStateBool(bool b)
+    : m_value(b)
 {
-    m_value = (b ? True : False);
 }
 
-TriStateBool::TriStateBool(TriStateBool::ValueType value)
-    : m_value(Undefined)
+TriStateBool TriStateBool::fromInt(int i)
 {
-    switch (value) {
-    case Undefined:
-    case True:
-    case False:
-        m_value = value;
+    switch (i) {
+    case -1: return TriStateBool::Undefined; break;
+    case 0: return TriStateBool::False; break;
+    case 1: return TriStateBool::True; break;
+    default:
+        qDebug() << Q_FUNC_INFO << "Incorrect value" << i;
+        return TriStateBool::Undefined;
     }
 }
 
-TriStateBool::operator bool() const
+TriStateBool TriStateBool::fromJsonValue(const QJsonValue &jsonVal)
 {
-    return (m_value == True);
+    if (jsonVal.isBool())
+        return jsonVal.toBool();
+
+    if (!jsonVal.isNull())
+        qDebug() << Q_FUNC_INFO << "Incorrect value" << jsonVal.toVariant();
```

Comment:
"Undefined" is not semantically equal to "incorrect" , but you seems to use "Undefined" in this way. All in all, I'm against this operator because there is no real conversion between a JSON value and bool value and it is illogical to pretend the other way.